tests/vector_tools/interpolate.cc

```diff
@@ -0,0 +1,282 @@
+
+
+  const unsigned int n_dofs_coarse = fe_system_coarse.dofs_per_cell;
+  const unsigned int n_dofs_fine   = fe_system_fine.dofs_per_cell;
+  const std::vector<dealii::Point<dim>> &points =
+    fe_system_fine.get_unit_support_points();
+
+  // Check that interpolated solution matches at the support points
+  std::vector<dealii::types::global_dof_index> dof_indices_coarse(
+    n_dofs_coarse);
+  std::vector<dealii::types::global_dof_index> dof_indices_fine(n_dofs_fine);
+
+  auto cell_coarse = dof_handler_coarse.begin_active();
+  auto cell_fine   = dof_handler_fine.begin_active();
+  auto endcell     = dof_handler_fine.end();
+
+  bool matching = true;
+  for (; cell_fine != dof_handler_fine.end(); ++cell_fine, ++cell_coarse)
+    {
+      if (!cell_fine->is_locally_owned())
+        continue;
+      cell_fine->get_dof_indices(dof_indices_fine);
+      cell_coarse->get_dof_indices(dof_indices_coarse);
+      for (unsigned int iquad = 0; iquad < points.size(); ++iquad)
+        {
+          dealii::Tensor<1, dim, double> local_solution_coarse;
+          dealii::Tensor<1, dim, double> local_solution_fine;
+          for (unsigned int idof = 0; idof < n_dofs_coarse; ++idof)
+            {
+              const unsigned int axis =
+                fe_system_coarse.system_to_component_index(idof).first;
+              local_solution_coarse[axis] +=
+                solution_coarse[dof_indices_coarse[idof]] *
+                fe_system_coarse.shape_value(idof, points[iquad]);
+            }
+          for (unsigned int idof = 0; idof < n_dofs_fine; ++idof)
+            {
+              const unsigned int axis =
+                fe_system_fine.system_to_component_index(idof).first;
+              local_solution_fine[axis] +=
+                solution_fine[dof_indices_fine[idof]] *
+                fe_system_fine.shape_value(idof, points[iquad]);
+            }
+          dealii::Tensor<1, dim, double> diff = local_solution_fine;
+          diff -= local_solution_coarse;
+          const double diff_norm = diff.norm();
+
+          std::cout << "Fine solution " << local_solution_fine
+                    << " Coarse solution " << local_solution_coarse
+                    << " Difference " << diff_norm << std::endl;
+          if (diff_norm > 1e-12)
+            matching = false;
+        }
+    }
+  if (matching)
+    dealii::deallog << "OK" << std::endl;
+  if (!matching)
+    dealii::deallog << "Non matching" << std::endl;
+}
+
+template <int dim>
+void
+run()
+{
+  using Vector = dealii::LinearAlgebra::distributed::Vector<double>;
+  // Create grid
+  dealii::parallel::distributed::Triangulation<dim> triangulation(
+    MPI_COMM_WORLD,
+    typename dealii::Triangulation<dim>::MeshSmoothing(
+      dealii::Triangulation<dim>::smoothing_on_refinement |
+      dealii::Triangulation<dim>::smoothing_on_coarsening));
+  const unsigned int n_1d_cells = 2;
+  dealii::GridGenerator::subdivided_hyper_cube(triangulation, n_1d_cells);
```

Comment:
Is the code here in essence duplicated from the 1d function, with the exception of using a `p::d::Triangulation`? If so, the easier approach might simply be to construct the triangulation in `main()` (non-distributed in 1d, distributed for 2d and 3d), and then pass it as an argument to this function. The body of the function would then be the same for 1d, 2d, and 3d.

Reply:
The duplicated code has been removed and the rest modified as suggested.